lib/synapse/service_watcher/base.rb

```diff
@@ -95,13 +100,44 @@ def validate_discovery_opts
     end
 
     def set_backends(new_backends)
-      if @keep_default_servers
-        @backends = @default_servers + new_backends
+      # Aggregate and deduplicate all potential backend service instances.
+      new_backends = (new_backends + (@keep_default_servers ? @default_servers : [])).uniq {|b|
+        [b['host'], b['port'], b.fetch('name', '')]
+      }
+
+      if new_backends.to_set == @backends.to_set
+        return false
+      end
+
+      if new_backends.empty?
+        if @default_servers.empty?
+          if @use_previous_backends
+            # Discard this update
+            log.warn "synapse: no default servers for service #{@name};" \
+              "  using previous backends: #{@backends.inspect}"
+            return false
+          else
+            log.warn "synapse: no default servers for service #{@name} and" \
+              " 'use_previous_backends' is disabled; dropping all backends"
+            @backends.clear
+          end
+        else
+          log.warn "synapse: no backends for service #{@name};" \
+            " using default servers: #{@default_servers.inspect}"
+          @backends = @default_servers
+        end
       else
+        log.info "synapse: discovered #{new_backends.length} backends for service #{@name}"
```

Comment:
I agree this is a needed feature. I've been hacking on a "latency sensitive farm failover" method as suggested in http://blog.haproxy.com/2013/12/23/failover-and-worst-case-management-with-haproxy/ and determining how many backends are needed is core to that.

The idea is that you have multiple HAProxy backends powering a single frontend, with each backend having more nodes (e.g. because it encompasses multiple datacenters).